analysis_driver/rest_communication.py

```diff
@@ -1,50 +1,68 @@
 from urllib.parse import urljoin
 import requests
-from pprint import pformat
 from analysis_driver.config import default as cfg
 from analysis_driver.app_logging import get_logger
 
 app_logger = get_logger(__name__)
 
 
-def api_url(endpoint):
-    return '{base_url}/{endpoint}/'.format(
+def api_url(endpoint, **query_args):
+    url = '{base_url}/{endpoint}/'.format(
         base_url=cfg.query('rest_api', 'url').rstrip('/'), endpoint=endpoint
     )
+    if query_args:
+        url += '?' + '&'.join(['%s=%s' % (k, v) for k, v in query_args.items()]).replace(' ', '').replace('\'', '"')
+
+    return url
 
 
 def _req(method, url, **kwargs):
     r = requests.request(method, url, **kwargs)
-    if r.status_code != 200:
-        app_logger.debug('%s %s %s %s' % (r.request.method, r.request.path_url, r.status_code, r.reason))
-        json = r.json()
-        if json:
-            app_logger.debug(pformat(json))
+    # e.g: 'POST <url> ({"some": "args"}) -> {"some": "content"}. Status code 201. Reason: CREATED
+    report = '%s %s (%s) -> %s. Status code %s. Reason: %s' % (
+        r.request.method, r.request.path_url, kwargs, r.content.decode('utf-8'), r.status_code, r.reason
+    )
+    if r.status_code in (200, 201):
+        app_logger.debug(report)
+    else:
+        app_logger.error(report)
     return r
 
+
+def depaginate_documents(endpoint, **queries):
+    elements = []
+    page_size = queries.pop('max_results', 100)
+    url = api_url(endpoint, max_results=page_size, **queries)
+    content = _req('GET', url).json()
+    elements.extend(content['data'])
+
+    if 'next' in content['_links']:
+        next_href, next_query = content['_links']['next']['href'].split('?')
+        next_query = dict([x.split('=') for x in next_query.split('&')])
+        queries.pop('page', None)
+        next_query.update(queries)
+        elements.extend(depaginate_documents(next_href, **next_query))
+    return elements
+
     
-def get_documents(endpoint, limit=10000, **kwargs):  # TODO: support embedding, max_results, etc.
-    url = api_url(endpoint)
-    param = []
-    url += '?max_results=%s' % limit
-    for key in kwargs:
-        param.append('"%s":"%s"' % (key, kwargs.get(key)))
-    if param:
-        url += '&where={%s}' % ','.join(param)
+def get_documents(endpoint, limit=10000, **query_args):
```

Comment:
If we're only allowing limited number of kwargs (where, embedded, and aggregate), shouldn't we specify them in get_document.
That would also help in knowing what paramter should be used.